docs/madr/decisions/019-offline-signing-tokens.md

```diff
@@ -0,0 +1,230 @@
+# Offline signing tokens
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4031
+
+## Context and Problem Statement
+
+Kuma provides multiple tokens for different lines of security
+* User authenticating to Kuma API Server
+* Data plane proxy authenticating to Kuma DP Server
+* Zone proxy authenticating to Kuma DP Server
+
+Right now, a user has to use Kuma CP to issue the tokens.
+While this works, some users may prefer using their existing token issuer.
+We can provide an interface of a token (document required fields) and a way to provide public key of a token.
+This way a user can use different system to issue token, but let Kuma CP validate the token.
+This improves security posture if you already have a safe way of managing signing key for your tokens.
+
+## Considered Options
+
+* CP configuration and kumactl improvements
+
+## Decision Outcome
+
+Chosen option: "CP configuration and kumactl improvements".
+
+## Pros and Cons of the Options
+
+### User token
+
+#### CP configuration
+
+```yaml
+apiServer:
+  authn:
+    type: tokens
+    tokens:
+      bootstrapAdminToken: true
+      enableIssuer: true
+      validator:
+        useSecrets: true
+        publicKeys:
+        - kid: 123
+          keyFile: /tmp/key.pem # either keyFile or key can be defined
+          key: |
+            -----BEGIN RSA PUBLIC KEY-----
+            MIIEogIBAAKCAQEAy0KtfI7O0TJ00............
+            -----END RSA PUBLIC KEY-----
+          mapping:
```

Comment:
Is it an arbitrary map, so that `mapping` has type `map[string]string`?

Reply:
I'm not sure if `oldName` and `newName` are good names. It's more of a `name of a field from your token`, `expected field`.

What I had in mind with "arbitrary map" is actually a typed object so
```go
type UserTokenMappings struct {
  Username string
  Groups string
}
```
maybe something like
```go
type UserTokenMappings struct {
  UsernameFieldName string
  GroupsFieldName string
}
```
would be more clear?

Reply:
What if a user wants other claims to be renamed, like using `meshName` instead of `mesh`? I think having this as a list would allow us to keep this API unchanged when we introduce new claims. 

We can rename it like:
```yaml
mapping:
  - expectedClaim: username
    tokenClaim: user
```

Reply:
> What if a user wants other claims to be renamed, like using meshName instead of mesh?

other claims? All possible claims would be in the `UserTokenMappings` object.

```
mapping:
  - expectedClaim: username
    tokenClaim: user
```
I'm ok with this API, mapping of claims will be a rare case anyway.

Reply:
Okay, if it's rare then probably doesn't really matter